### PR TITLE
Optimize nACL generation

### PR DESCRIPTION
**Pull request by YairSlobodin1**

### CODE REVIEW

File: pkg/synth/acl.go
Comment (zivnevo) [inline pkg/synth/acl.go:1]: Functions in this file have too many parameters. Spec and options are carried deep down the stack.
Consider a dedicated struct, and make all/some of the functions its methods. Something like
```golang
type ACLSynthesizer struct{
	spec *ir.Spec
	defs *ir.Definitions
	SingleACL bool
}
```

File: pkg/synth/acl.go [hunk 1]
```diff
@@ -15,153 +15,177 @@ import (
...
 		for _, dst := range conn.Dst.IPAddrs {
+			dstCidrs, dst := adjustResource(s, dst, conn.Dst)
 			if src == dst && conn.Src.Type != ir.ResourceTypeCidr && conn.Dst.Type != ir.ResourceTypeCidr {
 				continue
 			}
 			for _, trackedProtocol := range conn.TrackedProtocols {
 				reason := explanation{internal: internal, connectionOrigin: conn.Origin, protocolOrigin: trackedProtocol.Origin}
-				protocolRules := allowDirectedConnection(s, src, dst, conn, internalSrc, internalDst, trackedProtocol.Protocol, reason)
-				connectionRules = append(connectionRules, protocolRules...)
+				allowDirectedConnection(src, dst, srcCidrs, dstCidrs, internalSrc, internalDst, trackedProtocol.Protocol, reason, result, single)
```
Comment (zivnevo) [inline pkg/synth/acl.go:64]: The mapping of local variables to function parameters is confusing:
`src` -> `srcCidr`
`dst` -> `dstCidr`
`srcCidrs` -> `srcSubnets`
`dstCidrs` -> `dstSubnets`

File: pkg/synth/acl.go [hunk 2]
```diff
@@ -15,153 +15,177 @@ import (
...
-			acl.AppendExternal(rule)
-		}
-	}
 	return result
 }
 
-func allowDirectedConnection(s *ir.Spec, srcCidr, dstCidr *ipblock.IPBlock, conn *ir.Connection, internalSrc, internalDst bool,
-	protocol ir.Protocol, reason explanation) []*ir.ACLRule {
+// create relevant rule(s) to allow traffic
+func allowDirectedConnection(srcCidr, dstCidr *ipblock.IPBlock, srcSubnets, dstSubnets []*ir.ConnResource, internalSrc, internalDst bool,
```
Comment (zivnevo) [inline pkg/synth/acl.go:72]: I suggest this function is split into two functions: `allowConnectionFronSrc` and `allowConnectionToDst`

File: pkg/synth/acl.go [hunk 3]
```diff
@@ -15,153 +15,177 @@ import (
...
 		acl.AppendInternal(ir.DenyAllSend(subnet, cidr))
 	}
-
 	return result
 }
 
-func resourceRelevantToACL(e ir.ResourceType) bool {
-	return e == ir.ResourceTypeSubnet || e == ir.ResourceTypeCidr || e == ir.ResourceTypeNIF
+// Change ir.Resource to a slice of ir.ConnResource (each ir.ConnResource is either a pair of a of a subnet with its CIDR or an
+// external with its IPAddrs), to make it convenient to go through all pairs of subnets in allowDirectedConnection function.
```
Comment (zivnevo) [inline pkg/synth/acl.go:120]: Also explain what is the `addrs` parameter and why/when it is changed in the returned values.

File: pkg/synth/acl.go [hunk 4]
```diff
@@ -15,153 +15,177 @@ import (
...
 			}
-			request = &ir.Packet{Src: srcSubnetCidr, Dst: dstCidr, Protocol: protocol, Explanation: reason.String()}
-			connection = append(connection, ir.AllowSend(*request))
+			request = &ir.Packet{Src: srcSubnet.Addrs, Dst: dstCidr, Protocol: protocol, Explanation: reason.String()}
+			addRuleToACL(result, ir.AllowSend(*request), srcSubnet, internal, singleACL)
 			if inverseProtocol := protocol.InverseDirection(); inverseProtocol != nil {
-				response = &ir.Packet{Src: dstCidr, Dst: srcSubnetCidr, Protocol: inverseProtocol, Explanation: reason.response().String()}
-				connection = append(connection, ir.AllowReceive(*response))
+				response = &ir.Packet{Src: dstCidr, Dst: srcSubnet.Addrs, Protocol: inverseProtocol, Explanation: reason.response().String()}
+				addRuleToACL(result, ir.AllowReceive(*response), srcSubnet, internal, singleACL)
```
Comment (zivnevo) [inline pkg/synth/acl.go:86]: There is no point in taking the pointer of a struct and then only using it for dereferencing.
Better change `AllowReceive` and `AllowSend` (and `packetACLRule`) to take a pointer to `ir.Packet`.

File: pkg/synth/acl.go [hunk 5]
```diff
@@ -15,153 +15,177 @@ import (
...
+
+func adjustCidrSegment(s *ir.Definitions, cidr *ipblock.IPBlock, resourceName string) []*ir.ConnResource {
+	cidrSegmentDetails := s.CidrSegments[resourceName]
 	cidrDetails := cidrSegmentDetails.Cidrs[cidr]
-	result := make([]*ipblock.IPBlock, len(cidrDetails.ContainedSubnets))
+	result := make([]*ir.ConnResource, len(cidrDetails.ContainedSubnets))
 	for i, subnet := range cidrDetails.ContainedSubnets {
-		result[i] = s.Defs.Subnets[subnet].Address()
+		connResource := ir.ConnResource{Name: subnet, Addrs: s.Subnets[subnet].Address()}
+		result[i] = &connResource
```
Comment (zivnevo) [inline pkg/synth/acl.go:159]:
```suggestion
		result[i] = &ir.ConnResource{Name: subnet, Addrs: s.Subnets[subnet].Address()}
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/synth/acl.go [hunk 6]
```diff
@@ -15,153 +15,177 @@ import (
...
+
+func aclSelector(subnetName ir.ID, single bool) string {
+	if single {
+		return fmt.Sprintf("%s/singleACL", ir.VpcFromScopedResource(subnetName))
+	}
+	return subnetName
+}
+
+func resourceRelevantToACL(e ir.ResourceType) bool {
+	return e == ir.ResourceTypeSubnet || e == ir.ResourceTypeCidr || e == ir.ResourceTypeNIF || e == ir.ResourceTypeExternal
```
[the remark this answers is not in the document]
Comment (YairSlobodin1) [reply pkg/synth/acl.go:181]: `ResourceType` is determined by the endpoint type (e.g subnet segment is saved as a subnet)

File: cmd/subcmds/acl.go [hunk 7]
```diff
@@ -27,9 +27,9 @@ func NewACLCommand(args *inArgs) *cobra.Command {
 			}
 			var collection *ir.ACLCollection
 			if args.singleacl {
-				collection = synth.MakeACL(spec, synth.Options{SingleACL: true})
+				collection = synth.MakeACL(spec, true)
 			} else {
-				collection = synth.MakeACL(spec, synth.Options{SingleACL: false})
+				collection = synth.MakeACL(spec, false)
 			}
```
Comment (zivnevo) [inline cmd/subcmds/acl.go:33]:
```golang
			collection :=  synth.MakeACL(spec, args.singleacl)
```

File: pkg/ir/packet.go [hunk 8]
```diff
@@ -67,13 +67,13 @@ func makeDenyInternal() []ACLRule {
 
 func DenyAllSend(subnetName ID, cidr *ipblock.IPBlock) *ACLRule {
 	explanation := DenyAllExplanation(subnetName, cidr)
-	ACLRule := *packetACLRule(Packet{Src: cidr, Dst: ipblock.GetCidrAll(), Protocol: AnyProtocol{}, Explanation: explanation}, Outbound, Deny)
+	ACLRule := *packetACLRule(&Packet{Src: cidr, Dst: ipblock.GetCidrAll(), Protocol: AnyProtocol{}, Explanation: explanation}, Outbound, Deny)
 	return &ACLRule
```
Comment (zivnevo) [inline pkg/ir/packet.go:71]:
```suggestion
	return packetACLRule(&Packet{Src: cidr, Dst: ipblock.GetCidrAll(), Protocol: AnyProtocol{}, Explanation: explanation}, Outbound, Deny)
```

File: pkg/ir/packet.go [hunk 9]
```diff
@@ -67,13 +67,13 @@ func makeDenyInternal() []ACLRule {
...
 }
 
 func DenyAllReceive(subnetName ID, cidr *ipblock.IPBlock) *ACLRule {
 	explanation := DenyAllExplanation(subnetName, cidr)
-	ACLRule := *packetACLRule(Packet{Src: ipblock.GetCidrAll(), Dst: cidr, Protocol: AnyProtocol{}, Explanation: explanation}, Inbound, Deny)
+	ACLRule := *packetACLRule(&Packet{Src: ipblock.GetCidrAll(), Dst: cidr, Protocol: AnyProtocol{}, Explanation: explanation}, Inbound, Deny)
 	return &ACLRule
```
Comment (zivnevo) [inline pkg/ir/packet.go:77]:
```suggestion
	return packetACLRule(&Packet{Src: ipblock.GetCidrAll(), Dst: cidr, Protocol: AnyProtocol{}, Explanation: explanation}, Inbound, Deny)
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/synth/acl.go [hunk 10]
```diff
@@ -18,22 +18,20 @@ import (
...
-	spec      *ir.Spec
-	singleACL bool
-	result    *ir.ACLCollection
+	Spec      *ir.Spec
+	SingleACL bool
+	Result    *ir.ACLCollection
 }
 
-// MakeACL translates Spec to a collection of ACLs
+// MakeACL translatea.Spec to a collection of ACLs
```
Comment (zivnevo) [inline pkg/synth/acl.go:26]:
```suggestion
// MakeACL translates Spec to a collection of ACLs
```
OK to push this fix to the following PR

**Review by zivnevo — APPROVED** [review]

**Pull request merged**